geps/gep-1713/index.md

```diff
@@ -0,0 +1,380 @@
+
+eg.
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  infrastructure:
+    allowedChildren:
+      namespaces:
+        from: Same
+  gatewayClassName: example
+  listeners:
+  - name: HTTP
+    port: 80
+    protocol: HTTP
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    attachTo:
+      name: parent-gateway
+      kind: Gateway
+      group: gateway.networking.k8s.io
+  listeners:
+  - name: metrics
+    port: 8080
+    protocol: HTTP
+```
+
+With this configuration, the realized "parent" Gateway should listen on port `80` and `8080`.
+
+###### Listener Precedence
+
+Gateway Listeners should be merged using the following precedence:
+- "parent" Gateway
+- "child" Gateway ordered by creation time (oldest first)
+- "child" Gateway ordered alphabetical by “{namespace}/{name}”.
+
+If there are conflicts between these, this should be reported as `Conflicted=True` in the listener as usual.
+
+### Status Fields
+
+#### Addresses
+
+The list of `Addresses` that appear in the status of the "child" Gateway MUST be the same as the "parent" Gateway.
+
+#### Gateway Conditions
+
+Gateway conditions currently supports the following condition types: `Accepted` and `Programmed`
+
+For parent gateways, `Accepted` should be set based on the entire set of merged listeners.
+For instance, if a child listener is invalid, `ListenersNotValid` would be reported.
+`Programmed` is not expected, generally, to depend on the children resources, but if an implementation does depend on these
+they should consider child resources when reporting this status.
+
+For child gateways, `Accepted` and `Programmed` should not consider the overall merged Gateway status, but only the child's own listeners.
+The exception to this is if the parent Gateway is not valid in any regard.
+
+For example, if I have a `parent`, `child-1`, and `child-2`:
+* If `parent` is entirely invalid (for example, an invalid `address`), all three Gateways will reported `Accepted=False`.
+* If `child-1` has an invalid listener, `parent` and `child-1` will report `ListenersNotValid`, while `child-2` will not.
+* If `child-1` references a parent that doesn't exist then `child-1` will report `Accepted=False`
+* If `child-1` references a parent that doesn't allow merging then `child-1` will report `Accepted=False`
+* If `child-1` references another child (eg. `child-2`) then `child-1` will report `Accepted=False` 
+* If `child-1` references itself then `child-1` will report `Accepted=False`
+* If `child-1` and `parent` have different gatewayClassNames then `child-1` will report `Accepted=False`
+* If `child-1` references a parent whose allowed child namespaces do not include `child-1`'s namespace, then `child-1` will report `Accepted=False`
```

Comment:
One other scenario here:

`parent`, `child-1`, `child-2`, and `child-3`.

All three children have multiple listeners.
`child-2` and `child-3` have _one_ listener that conflicts.
Let's say `child-2` was created first.

What happens?

How do the owners of each child know what's happened? (Preferably without looking at the parent's status)
What actions do the owners of `child-2` and `child-3` need to take to fix this? Can they know that which other sibling Gateway has caused the problem? If `child-2` wins precedence because of age, how do they know that something tried to claim their config and failed?

Note that in this case, if `child-2` ever deletes and recreates their Gateway, they will now be in the same position as the owner of `child-3`.

How does the owner of the parent know if any children are bad, and is it possible for them to know which ones?

Reply:
> What happens?

`child-3` would be rejected and `child-2` wouldn't change. I'll add a condition about said conflict - I thought this was covered elsewhere.

> How do the owners of each child know what's happened? (Preferably without looking at the parent's status)

The status on the child could state an appropriate error message.

> What actions do the owners of child-2 and child-3 need to take to fix this? 

Gateway owners (cluster operator) would have to address this by tweaking listeners.

> Can they know that which other sibling Gateway has caused the problem? 

I think it's worth the error message saying there's a conflict and providing the more detailed message about siblings on the parent's status message.

> If child-2 wins precedence because of age, how do they know that something tried to claim their config and failed? 

Do you have a use case why child-2 would _want_ to know this information?

> Note that in this case, if child-2 ever deletes and recreates their Gateway, they will now be in the same position as the owner of child-3.

Yup

> How does the owner of the parent know if any children are bad, and is it possible for them to know which ones?

It's mentioned that `ListenersNotValid` will appear on the parent's status. A more detailed message can be presented.